File: packages/npm/src/utils/sentinel-merge.ts
import { outputFile, pathExists } from 'fs-extra'
import { lstat, readFile, writeFile } from 'node:fs/promises'
import { basename, dirname } from 'node:path'
import { writeBlocked } from './fs-safe.js'

const SENTINEL_START = '<!-- goodvibes:start -->'
const SENTINEL_END = '<!-- goodvibes:end -->'
// A marker counts only alone on its line; the lookahead keeps trailing whitespace and \r out of the match.
const START_LINE = /^<!-- goodvibes:start -->(?=[ \t\r]*$)/gm
const END_LINE = /^<!-- goodvibes:end -->(?=[ \t\r]*$)/gm

const TAG = '(?:alpha|beta|rc|a|b|post|dev)'

export function extractVersion(block: string): string | null {
  const match = block.match(new RegExp(`# goodvibes: v(\\d+(?:\\.\\d+)*(?:[-.]?${TAG}(?:\\.?\\d+)?\\b)?)`))
  return match ? match[1] : null
}

const PHASE: Record<string, number> = { dev: -4, alpha: -3, a: -3, beta: -2, b: -2, rc: -1, post: 1 }

function parseVersion(v: string): number[] | null {
  const m = v.trim().match(new RegExp(`^v?(\\d+)(?:\\.(\\d+))?(?:\\.(\\d+))?(?:[-.]?(${TAG})(?:\\.?(\\d+))?)?$`, 'i'))
  if (!m) return null
  return [Number(m[1]), Number(m[2] ?? 0), Number(m[3] ?? 0), m[4] ? PHASE[m[4].toLowerCase()] : 0, Number(m[5] ?? 0)]
}

// Unparseable input compares as "not newer", so callers never skip an update or install on a bad string.
export function versionGte(a: string, b: string): boolean {
  const pa = parseVersion(a)
  const pb = parseVersion(b)
  if (!pa || !pb) return false
  for (let i = 0; i < pa.length; i++) {
    if (pa[i] !== pb[i]) return pa[i] > pb[i]
  }
  return true
}

function extractSentinelBlock(content: string): string {
  const start = content.indexOf(SENTINEL_START)
  const end = content.indexOf(SENTINEL_END)
  if (start === -1 || end === -1) return ''
  return content.slice(start, end + SENTINEL_END.length)
}

export class MarkerError extends Error {}

function markerProblem(starts: number[], ends: number[]): string | null {
  if (starts.length > 1) return `has ${starts.length} ${SENTINEL_START} lines; expected one`
  if (ends.length > 1) return `has ${ends.length} ${SENTINEL_END} lines; expected one`
  if (starts.length === 1 && ends.length === 0) return `has a ${SENTINEL_START} line but no ${SENTINEL_END} line`
  if (ends.length === 1 && starts.length === 0) return `has a ${SENTINEL_END} line but no ${SENTINEL_START} line`
  if (starts.length === 1 && ends[0] < starts[0]) return `has its ${SENTINEL_END} line before its ${SENTINEL_START} line (the end line comes before the start line)`
  return null
}

// Checked here, just before the write: a caller's earlier check can be stale by the time the file is written.
async function refuseLink(destPath: string): Promise<void> {
  if (!(await lstat(dirname(destPath)).catch(() => null))) return // mergeClaude may create the folder; a missing folder is no link
  const why = await writeBlocked(dirname(destPath), basename(destPath))
  if (why) throw new MarkerError(why)
}

// Node's plain utf-8 read turns a bad byte into U+FFFD, which the write would then save over the user's text.
async function readUtf8(destPath: string): Promise<string> {
  try {
    return new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(await readFile(destPath))
  } catch (e) {
    if (!(e instanceof TypeError)) throw e
    throw new MarkerError(`${destPath} is not UTF-8 text, so goodvibes did not change it; save it as UTF-8 and re-run.`)
  }
}

function markers(destPath: string, existing: string): { starts: number[]; ends: number[] } {
  const starts = [...existing.matchAll(START_LINE)].map(m => m.index)
  const ends = [...existing.matchAll(END_LINE)].map(m => m.index)
  const problem = markerProblem(starts, ends)
  if (problem) {
    throw new MarkerError(
      `${destPath} ${problem}, so goodvibes did not change it; fix CLAUDE.md by hand: keep exactly one ${SENTINEL_START} line ` +
        `followed later by one ${SENTINEL_END} line, or delete both to get a fresh block.`,
    )
  }
  return { starts, ends }
}

// Removes the goodvibes block and keeps the text around it; true when there was a block.
export async function stripBlock(destPath: string, dryRun = false): Promise<boolean> {
  await refuseLink(destPath)
  if (!(await pathExists(destPath))) return false
  const existing = await readUtf8(destPath)
  const { starts, ends } = markers(destPath, existing)
  if (starts.length === 0) return false
  if (!dryRun) {
    const eol = existing.includes('\r\n') ? '\r\n' : '\n'
    const after = existing.slice(ends[0] + SENTINEL_END.length).replace(/^(?:[ \t]*\r?\n)+/, '')
    const parts = [existing.slice(0, starts[0]).trimEnd(), after.trimEnd()].filter(Boolean)
    await writeFile(destPath, parts.join(eol + eol) + (parts.length > 0 ? eol : ''))
  }
  return true
}

// Throws MarkerError without writing when the markers are ambiguous, so no user text is ever cut.
export async function mergeClaude(destPath: string, templateContent: string): Promise<void> {
  const templateBlock = extractSentinelBlock(templateContent)
  await refuseLink(destPath)

  if (!(await pathExists(destPath))) {
    await outputFile(destPath, templateContent)
    return
  }

  const existing = await readUtf8(destPath)
  const eol = existing.includes('\r\n') ? '\r\n' : '\n'
  const block = templateBlock.replace(/\r?\n/g, eol)
  const { starts, ends } = markers(destPath, existing)

  if (starts.length === 0) {
    await writeFile(destPath, existing.trimEnd() + eol + eol + block + eol)
    return
  }

  const startIdx = starts[0]
  const endIdx = ends[0] + SENTINEL_END.length
  const existingVersion = extractVersion(existing.slice(startIdx, endIdx))
  const templateVersion = extractVersion(templateBlock)
  if (existingVersion && templateVersion && versionGte(existingVersion, templateVersion)) return

  await writeFile(destPath, existing.slice(0, startIdx) + block + existing.slice(endIdx))
}

File: packages/pip/src/goodvibes_cli/utils/sentinel_merge.py
"""CLAUDE.md sentinel block merge — stdlib only port of sentinel-merge.ts."""
from __future__ import annotations

import re
import pathlib

SENTINEL_START = "<!-- goodvibes:start -->"
SENTINEL_END = "<!-- goodvibes:end -->"


_VERSION = r"\d+(?:\.\d+)*(?:[-.]?(?:alpha|beta|a|b|rc)\.?\d*)?(?:\.post\d+)?"
_PARTS = re.compile(r"v?(\d+(?:\.\d+)*)(?:[-.]?(alpha|beta|a|b|rc)\.?(\d*))?(?:\.post(\d+))?\.?", re.I)
_PRE_RANK = {"alpha": 0, "a": 0, "beta": 1, "b": 1, "rc": 2}


def extract_version(block: str) -> str | None:
    """Return version string from a sentinel block or version stamp line, or None."""
    m = re.search(rf"# goodvibes: v({_VERSION})", block)
    return m.group(1) if m else None


def _version_key(v: str) -> tuple | None:
    m = _PARTS.fullmatch(v.strip())
    if not m:
        return None
    release = [int(x) for x in m.group(1).split(".")]
    while len(release) > 1 and release[-1] == 0:
        release.pop()
    # A release sorts above its pre-releases (rank 3) and below its .postN releases.
    pre = (_PRE_RANK[m.group(2).lower()], int(m.group(3) or 0)) if m.group(2) else (3, 0)
    post = int(m.group(4)) + 1 if m.group(4) is not None else 0
    return (release, pre, post)


def version_gte(a: str, b: str) -> bool:
    """Return True if version a >= version b; False when either cannot be parsed."""
    ka, kb = _version_key(a or ""), _version_key(b or "")
    if ka is None or kb is None:
        return False
    return ka >= kb


def _extract_sentinel_block(content: str) -> str:
    start = content.find(SENTINEL_START)
    end = content.find(SENTINEL_END)
    if start == -1 or end == -1:
        return ""
    return content[start : end + len(SENTINEL_END)]


class ClaudeMdError(ValueError):
    pass


# A marker counts only alone on its line, so prose that mentions a marker is never treated as one.
_START_LINE = re.compile(rf"^{re.escape(SENTINEL_START)}[ \t]*(?=\r?$)", re.M)
_END_LINE = re.compile(rf"^{re.escape(SENTINEL_END)}[ \t]*(?=\r?$)", re.M)


def _read_text(path: pathlib.Path) -> str:
    try:
        with open(path, encoding="utf-8", newline="") as f:
            return f.read()
    except UnicodeDecodeError as e:
        raise ClaudeMdError(f"{path.name}: not UTF-8 text (byte {e.start}); goodvibes left it unchanged. Save it as UTF-8 and re-run.") from e


def _write_text(path: pathlib.Path, text: str) -> None:
    with open(path, "w", encoding="utf-8", newline="") as f:
        f.write(text)


def _marker_problem(starts: list, ends: list) -> str | None:
    if len(starts) > 1:
        return f"it has {len(starts)} {SENTINEL_START} lines"
    if len(ends) > 1:
        return f"it has {len(ends)} {SENTINEL_END} lines"
    if starts and not ends:
        return f"it has a {SENTINEL_START} line but no {SENTINEL_END} line"
    if ends and not starts:
        return f"it has a {SENTINEL_END} line but no {SENTINEL_START} line"
    if ends[0].start() < starts[0].start():
        return f"its {SENTINEL_END} line comes before its {SENTINEL_START} line"
    return None


def merge_claude(dest_path: pathlib.Path, template_content: str) -> None:
    """Merge goodvibes sentinel block into dest_path from template_content.

    Case A: dest absent → write template verbatim.
    Case B: dest exists, no marker lines → append sentinel block.
    Case C: dest exists, one start line then one end line, older version → replace between them.
    Case D: dest exists, sentinel same or newer → no write.
    Any other marker layout raises ClaudeMdError without writing.
    """
    template_block = _extract_sentinel_block(template_content)

    if not dest_path.exists():
        # Case A
        dest_path.parent.mkdir(parents=True, exist_ok=True)
        _write_text(dest_path, template_content)
        return

    existing = _read_text(dest_path)
    nl = "\r\n" if "\r\n" in existing else "\n"
    block = template_block.replace("\r\n", "\n").replace("\n", nl)
    starts, ends = list(_START_LINE.finditer(existing)), list(_END_LINE.finditer(existing))

    if not starts and not ends:
        # Case B: no sentinel — append block
        _write_text(dest_path, existing.rstrip() + nl + nl + block + nl)
        return

    problem = _marker_problem(starts, ends)
    if problem:
        raise ClaudeMdError(
            f"{dest_path.name}: {problem}; goodvibes left it unchanged. Please fix CLAUDE.md by hand: keep exactly one "
            f"{SENTINEL_START} line followed later by one {SENTINEL_END} line (or delete both), then re-run."
        )

    start_idx, end_idx = starts[0].start(), ends[0].end()
    existing_version = extract_version(existing[start_idx:end_idx])
    template_version = extract_version(template_block)

    if existing_version and template_version and version_gte(existing_version, template_version):
        # Case D: existing version >= template — skip
        return

    # Case C: replace only sentinel block
    _write_text(dest_path, existing[:start_idx] + block + existing[end_idx:])
